File: lua/test/test_log_tracer.lua
local d = require("mempeep.descriptors")
local log_tracer = require("mempeep.tracers.log_tracer")
local memory = require("mempeep.test.memory")
local read = require("mempeep.read")

local mock_out = function(text)
  local lines = {}
  for line in text:gmatch("[^\n]+") do
    table.insert(lines, line)
  end
  local out = {}
  function out:write(s)
    for line in s:gmatch("[^\n]+") do
      assert(line == lines[1], "expected '" .. lines[1] .. "' but got '" .. s .. "'")
    end
    table.remove(lines, 1)
  end
  return out
end

do
  local Inner = d.Struct("Inner", { d.Field(d.Int32, "a"), d.Field(d.Int32, "b") })
  local Outer = d.Struct("Outer", { d.Field(Inner, "inner"), d.Field(d.Int32, "c") })
  local reader = memory.mock_memory_reader(
    "I4",
    "\x0B\x00\x00\x00" -- a = 11
      .. "\x16\x00\x00\x00" -- b = 22
      .. "\x21\x00\x00\x00" -- c = 33
  )
  local out = mock_out([[
[00000000] .inner.a = 0xb
[00000004] .inner.b = 0x16
[00000008] .c = 0x21
]])
  local tracer = log_tracer.new(log_tracer.on_entry_write(out), log_tracer.log_level.VALUES)
  local v, ok = read.read(Outer, 0, reader, tracer)
  assert(ok)
  assert(v)
  assert(v.inner.a == 11)
  assert(v.inner.b == 22)
  assert(v.c == 33)
end

File: cpp/test/test_log_tracer.cpp
#define DOCTEST_CONFIG_IMPLEMENT_WITH_MAIN
#include <doctest/doctest.h>

#include <mempeep/read.hpp>
#include <mempeep/test/memory.hpp>
#include <mempeep/tracers/log_tracer.hpp>

using namespace mempeep;

using TInt32 = Primitive<int32_t>;

struct Inner {
  int32_t a;
  int32_t b;
};

using TInner
  = Struct<Inner, Fields<Field<TInt32, &Inner::a>, Field<TInt32, &Inner::b>>>;

struct Outer {
  Inner inner;
  int32_t c;
};

using TOuter = Struct<
  Outer,
  Fields<Field<TInner, &Outer::inner>, Field<TInt32, &Outer::c>>>;

TEST_CASE("simple") {
  auto reader = test::MockMemoryReader<uint8_t>{
    "\x0B\x00\x00\x00\x16\x00\x00\x00\x21\x00\x00\x00"
  };
  Outer outer{};
  std::stringstream out;
  auto tracer = LogTracer{on_entry_print(out), LogLevel::VALUES};
  CHECK(read<TOuter>(0, reader, tracer, outer));
  CHECK_EQ(out.str(), R"([00000000] .inner.a = 0xb
[00000004] .inner.b = 0x16
[00000008] .c = 0x21
)");
  CHECK_EQ(outer.inner.a, 0x0b);
  CHECK_EQ(outer.inner.b, 0x16);
  CHECK_EQ(outer.c, 0x21);
  CHECK(tracer.success());
}
